enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -182,4 +191,72 @@ impl<'a> Handler<'a> {
             ret
         }
     }
+
+    /// Do a readv() syscall
+    pub fn readv(&mut self) -> u64 {
+        let fd = self.aex.gpr.rdi;
+        let trusted = unsafe {
+            from_raw_parts_mut(self.aex.gpr.rsi as *mut Iovec, self.aex.gpr.rdx as usize)
+        };
+
+        // Add up total size of buffers.
+        let bufsize = trusted.iter().fold(0, |a, e| a + e.size);
+
+        // Allocate some unencrypted memory.
+        let size = bufsize + size_of::<Iovec>() * trusted.len();
+        let map = match self.ualloc(size as u64) {
+            Err(errno) => return errno.into_syscall(),
+            Ok(map) => unsafe { from_raw_parts_mut(map, size as usize) },
+        };
+
+        // Split allocated memory into that used by the iovec struct array and that used by its buffers.
+        let (uiovec, ubuffer) = map.split_at_mut(size_of::<Iovec>() * trusted.len());
+
+        // Convert the prefix from a byte slice into an iovec slice.
+        let (_, untrusted, _) = unsafe { uiovec.align_to_mut::<Iovec>() };
+        if untrusted.len() != trusted.len() {
+            self.attacked();
+        }
+
+        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
+        // The offset is into the buffer area allocated immediately after the iovec struct
+        // array, measured in bytes.
+        let mut offset = 0;
+        for (t, mut u) in trusted.iter_mut().zip(untrusted.iter_mut()) {
+            u.base = ubuffer[offset] as *mut u8;
+            u.size = t.size;
+            offset += t.size;
+        }
+
+        unsafe {
+            // Do the syscall; replace encrypted memory with unencrypted memory.
+            let ret = self.syscall(
+                SysCall::READV,
+                fd,
+                untrusted.as_ptr() as _,
+                trusted.len() as u64,
+                0,
+                0,
+                0,
+            );
+
+            // Copy the unencrypted input into encrypted memory.
+            if ErrNo::from_syscall(ret).is_none() {
+                if ret > size as u64 {
+                    self.attacked();
+                }
+
+                let mut offset = 0;
+                for (t, u) in trusted.iter().zip(untrusted.iter()) {
+                    if u.base != ubuffer[offset] as *mut u8 || u.size != t.size {
+                        self.attacked();
+                    }
+                    core::ptr::copy_nonoverlapping(u.base as *mut u8, t.base as *mut u8, ret as _);
```

Comment:
Use the slice method `copy_from_slice()` instead.

Reply:
Since `u.base` is a `*mut u8`, would I not have to turn it into a slice with the unsafe `core::slice::from_raw_parts_mut` in order to use that method? That would result in the same number `unsafe` lines here. Is there a better way?

Reply:
You know the offset into the `ubuffer`, which is already a slice.